### PR TITLE
Cannibals are pariahs 2

### PR DESCRIPTION
**Pull request by RenechCDDA**
#### Summary
None

#### Purpose of change
Followup to #77793

I only flagged the beggars as getting mad if you force fed them. People pointed out that this was insufficient. I agree.

Any signs of cannibalism will be treated with ostracizing at best.

#### Describe the solution
-Butchering people for meat causes anyone that can see you and is not okay with it to take massive opinion hits

-Selling/giving cannibal products to the refugee center will result in exile. You're given a minute to, essentially, run the hell away.

-Food with cannibalism vitamins is worthless

-Non-cannibal/psychopath/sapiovore NPCs will refuse to butcher human corpses. 

I extracted a common check of being a canibal/psychopath/sapiovore into a common Character::okay_with_eating_humans() function.

I had to expand json's ability to access faction_food_supply to get/set vitamins

#### Describe alternatives you've considered
I would like to have Smokes' confrontation to be forced and immediate next time you're at the refugee center, but it is mechanically 1000x simpler and easier to maintain to only do it when the trading topic is selected.

#### Testing
Cannibal food is worth 0
image

Selling the refugee center cannibal food causes Smokes to confront you next time you bring up trading. (Note that this example used modified vitamin values because of https://github.com/CleverRaven/Cataclysm-DDA/pull/77825#issuecomment-2481582960)

https://github.com/user-attachments/assets/b9edad3d-affa-4463-801d-d686f7e8f319

Butchering people in front of those who are not okay with it results in them getting angry.

image

image

#### Additional context
Known issues:
NPCs that are not okay with doing butchering will still "help [you] with this task..."

There's not a simple way to untangle it, since that message is given *before the activity even begins*, let alone before you are prompted to actually confirm your intentions.

### CODE REVIEW

**Comment by Naadn** [comment]
I have a question : How do other people know the mysterious meat randos sell to them is human meat?

**Comment by SariusSkelrets** [comment]
I wonder that too, it's not like if they regularly see human flesh or can know at a glance the difference between cooking oil made from human fat and cooking oil made from non-human fat.

**Comment by RenechCDDA** [comment]
The same way you do. This has already been discussed, so please refer to the previous PR instead of bringing it up again.

**Comment by ampersand55** [comment]
I like that there are cannibal aspects in the game.

(Bare with me for a moment, I'm not an edge lord)

Take a moment to consider the implications of the fact that a *starving* beggar camping outside the refugee camp can instantly recognize human meat... That's some horrific immersive story-telling, world-building and character development all in one, from just a simply game mechanic! It ties in so well with the cataclysm setting of a growing sense of desperation and dread. 

Lean into it, I would welcome the refugee center to prominently taking a hard stance against cannibalism. Have them attack you on sight. Maybe a quests to kill a band of sapiovores, or a revenge quest from the beggar? Perhaps there could be some other faction that you slowly learn are actually cannibals?

**Comment by kevingranade** [comment]
As outlined in the previous PR, we are not interested in "leaning into" cannibalism-related content.  The current stance is "yes it's possible but here are the repercussions of doing so".

**Comment by IdleSol** [comment]
What about marlo and turning beggars into mushrooms? I.e. does it work at all? And if it works, shouldn't this patch be extended to marlo as well?

**Comment by Bone-blood-moss** [comment]
Just make sure your follows don't care. Or it will 100% begin Considered a bug And add more reason to not have followers.

**Comment by RenechCDDA** [comment]
> Just make sure your follows don't care. Or it will 100% begin Considered a bug And add more reason to not have followers.

There is no reason for your followers to suddenly become mindless drones.

**Comment by ampersand55** [comment]
Perhaps some `has_trait` check for cannibal, strict humanitarian or hunter/predator/apex predator.

**Comment by RenechCDDA** [comment]
> Perhaps a `has_trait` check on npcs for cannibal, strict humanitarian or hunter/predator/apex predator to see how they react to a cannibal player.

That was the planned implementation of Character::okay_with_eating_people(). You'll note it's a dummy (non-existent) at the time of this writing

**Comment by RenechCDDA** [comment]
The refugee center's protections will not be fully effective in this PR. It's way out of scope so I plan to make a followup for this.

image

**Comment by RenechCDDA** [comment]
> [...] And if the NPCs somehow magically know about the composition. Then they should react. [...]

Yes, I agree, they should be similarly upset if you give them mutant meat, or something disgusting or (basically) inedible. But that's not within the scope of this PR. This PR is aimed at cannibalism and nothing else.

File: src/math_parser_diag.cpp [hunk 1]
```diff
@@ -268,19 +268,28 @@ diag_assign_dbl_f faction_trust_ass( char /* scope */, std::vector<diag_value> c
...
 diag_eval_dbl_f faction_food_supply_eval( char /* scope */,
-        std::vector<diag_value> const &params, diag_kwargs const &/* kwargs */ )
+        std::vector<diag_value> const &params, diag_kwargs const &kwargs )
 {
-    return [fac_val = params[0]]( const_dialogue const & d ) {
+    diag_value vit_val = kwargs.kwarg_or( "vitamin" );
+    return [fac_val = params[0], vit_val]( const_dialogue const & d ) {
         faction *fac = g->faction_manager_ptr->get( faction_id( fac_val.str( d ) ) );
+        if( !vit_val.is_empty() ) {
+            return static_cast<double>( fac->food_supply.get_vitamin( vitamin_id( vit_val.str() ) ) );
```
[the remark this answers is not in the document]
Comment (RenechCDDA) [reply src/math_parser_diag.cpp:277]: Good catch

**Comment by Maleclypse** [comment]
@RenechCDDA  can you resolve the conflict and ping me when this is  basic build passed?

**Comment by Bone-blood-moss** [comment]
If you do add Mutant meat make sure there's a timer for it. It should goes off after some time cause animals will mutate more and more. Not sure Humanoid/demhuman meat. Should be trust based for when we add nether meat.

**Comment by RenechCDDA** [comment]
> @RenechCDDA can you resolve the conflict and ping me when this is basic build passed?

I think it still needs something to address the points @Andrei8l brought up.



> If you do add Mutant meat make sure there's a timer for it. It should goes off after some time cause animals will mutate more and more. Not sure about Humanoid/demhuman meat. Should be trust based for when we add nether meat.

This PR only address cannibalism.

**Comment by RenechCDDA** [comment]
On later thought I'm not getting anywhere with it, so that can be a followup.

**Pull request merged**